File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        inDegrees, outgoing = [0] * numCourses, defaultdict(set)
        
        for course, pre in prerequisites:
            outgoing[pre].add(course)
            inDegrees[course] += 1

        q, count = deque(), 0
        for i in range(numCourses):
            if inDegrees[i] == 0:
                q.append(i)
        
        while q:
            course = q.popleft()
            count += 1
            for neighbor in outgoing[course]:
                inDegrees[neighbor] -= 1
                if inDegrees[neighbor] == 0:
                    q.append(neighbor)
                    
        return count == numCourses

File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    vector<int>visited;
    map<int,vector<int>>adj;
    
    bool dfs(int v) {
        if (visited[v] == 1)
        return false;
        if (visited[v] == 2)
            return true;
        visited[v] = 1;
        for (auto u: adj[v]) {
            if (!dfs(u))
                return false;
        }
        visited[v] = 2;
        return true;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        visited.assign(numCourses, false);
        for (auto p: prerequisites)
            adj[p[1]].push_back(p[0]);
        
        for (int i = 0; i < numCourses; i++)
            if (!dfs(i))
                return false;
        return true;
    }
};